### ISSUE TITLE
Check regexes at compile

### ISSUE THREAD

**Issue by gabriel-fallen** [typeable/schematic#8]
Currently regular expressions (`TRegex` for `SchemaText`) get compiled only when (and if) relevant data being validated at run-time. Thus even malformed regexes pass compilation and might sleep past testing. It would be nice if they get compiled or syntactically verified while compiling Haskell.

**Comment by dredozubov** [comment]
It would. But do you have a way of doing that in mind?

**Comment by sboosali** [comment]
There's not much you can do with type-level strings even in GHC 8.2.
There's probably a type checker plug in somewhere, or a macro would work,
but those have the obvious drawbacks.

**Comment by sboosali** [comment]
However, a type-level regex representation, well it's convenient, could
work.

Like a data kind for each case:

data REGEX = LITERAL Symbol | STAR REGEX | ...

**Comment by dredozubov** [comment]
I've thought about the type level representation, but I would rather separate something like this to a different project before trying to incorporate it into schematic. Schematic already incurs heavy compilation burden on ghc and I think it can get worse with type level regexes. Do you think some helpers to test out the type level regexes would come in handy? It may be an alternative.

**Comment by sboosali** [comment]
Yeah, wrapping whatever regular expression parser is eventually invoked makes sense.
